libgnucash/core-utils/test/gtest-gnc-unicode.cpp

```diff
@@ -0,0 +1,135 @@
+ * published by the Free Software Foundation; either version 2 of   *
+ * the License, or (at your option) any later version.              *
+ *                                                                  *
+ * This program is distributed in the hope that it will be useful,  *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of   *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
+ * GNU General Public License for more details.                     *
+ *                                                                  *
+ * You should have received a copy of the GNU General Public License*
+ * along with this program; if not, contact:                        *
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+ *                                                                  *
+\********************************************************************/
+
+#include <gnc-unicode.h>
+#include <stdbool.h>
+#include <gtest/gtest.h>
+
+TEST(GncUnicode, test_ss_primary)
+{
+    int pos = 0, len = 0;
+
+    auto result = gnc_unicode_has_substring_base("besi", "Necklace for Bessie",
+                                         &pos, &len);
+    EXPECT_FALSE(result);
+    EXPECT_EQ(0, pos);
+    EXPECT_EQ(0, len);
+
+    result = gnc_unicode_has_substring_base("bessi", "Necklace for Bessie",
+                                         &pos, &len);
+    EXPECT_TRUE(result);
+    EXPECT_EQ(13, pos);
+    EXPECT_EQ(5, len);
+    }
+TEST(GncUnicode, test_ss_secondary)
+{
+    int pos = 0, len = 0;
+    auto result = gnc_unicode_has_substring_accented_case_insensitive("bessi", "Necklace for Bessie",
+                                         &pos, &len);
+    EXPECT_TRUE(result);
+    EXPECT_EQ(13, pos);
+    EXPECT_EQ(5, len);
+}
+
+TEST(GncUnicode, test_ss_secondary_case_sensitive)
+{
+    int pos = 0, len = 0;
+    auto result = gnc_unicode_has_substring_accented_case_sensitive("bessi", "Necklace for Bessie",
+                                         &pos, &len);
+    EXPECT_FALSE(result);
+    EXPECT_EQ(0, pos);
+    EXPECT_EQ(0, len);
+
+    result = gnc_unicode_has_substring_accented_case_sensitive("Bessi", "Necklace for Bessie",
+                                         &pos, &len);
+    EXPECT_TRUE(result);
+    EXPECT_EQ(13, pos);
+    EXPECT_EQ(5, len);
+}
+
+TEST(GncUnicode, test_german_ss_literal)
+{
+    int pos = 0, len = 0;
+    auto result = gnc_unicode_has_substring_base("be\xc3\x9fi", "Necklace for Be\xc3\x9fie",
+                                         &pos, &len);
+    EXPECT_TRUE(result);
+    EXPECT_EQ(13, pos);
+    EXPECT_EQ(4, len);
+
+    pos = 0, len = 0;
+    result = gnc_unicode_has_substring_accented_case_insensitive("be\xc3\x9fi", "Necklace for Be\xc3\x9fie",
```

Comment:
Out of curiosity, is it bad practice to put the actual unicode character directly in a string constant in C++ ? Don't we have [string literal qualifiers](http://www.en.cppreference.com/w/cpp/language/string_literal.html) now that help express the string is supposed to be interpreted as unicode ?

For example isn't `u8"Be쎟ie"` equivalent to `"Be\xc3\x9fi"` ?

Reply:
Search the web for "utf-8 c3 9f" or use any of the myriad utf-8 calculator websites. The [details](https://en.wikipedia.org/wiki/UTF-8#Description) in case you're interested.

Reply:
You're correct about my misinterpretation as UTF-16. My confusion comes from the Character map tool on my linux system that shows these UTF-16 code points for each symbol. I didn't realize that. Looking better I find it also displays UTF-8 further down on the info panel, but I can't search on it...